### DISCUSSION TITLE
Component: Token Icon (Wallet + Bridge)

### DISCUSSION THREAD

**Discussion by heycorwin** [near/near-wallet#1530] [Show and tell]
## Problem Summary
The wallet and bridge teams are in need of a standard for how to display token icons. With the addition of `icon` in the [token contract metadata](https://github.com/near/near-sdk-rs/pull/315), we now have an option for token contracts to support an icon. With its addition, we need to consider the following:

1. A "default" or "empty state" icon, that is displayed when no icon is provided by the contract.
2. An icon "standard" to inform how token icons should be displayed in the context of a UI
3. A way to indicated modified tokens (eg wrapped or bridged) so that they may be distinguishable from unmodified tokens (eg vanilla NEAR or an unmodified NEP141)

**This discussion is related to:**
- [Display of Fungible Token Balances](https://github.com/near/near-wallet/issues/1147)
- [Display of bridged tokens within the bridge UI](https://github.com/near/rainbow-bridge-frontend/issues/117)

## Solution
The following guideline is an attempt to outline a simple but flexible system to display token icons in the context of an interface. My suggestion is that we create a component that takes the icon provided by the contract (if provided), and displays it in various states depending on if and how the token has been modified.

![Token Icon Component](https://user-images.githubusercontent.com/5934188/110366878-e5eabe80-8014-11eb-9f26-71cfd40166eb.jpg)

cc @chadoh @djsatok @vgrichina @Patrick1904

**Comment by stefanopepe** [comment]
To cut time to release a token, is it possible to also use unicode to generate the icon? https://unicode-table.com/en/blocks/
This would instantly enable a very large set of characters, with the cost of a few bytes (up to 12)

**Comment by alexauroradev** [comment]
Looks nice, but what about some services that rounds the icons, like https://coinmarketcap.com/ ? Prepended modifier may not be shown correctly.

**Comment by heycorwin** [comment]
Alright so @djsatok made a great point that third party exchanges are going to have limitations of their own for how they choose to display token icons. It's in our best interest to avoid any scenarios where any visual elements from our approach might get clipped. The standard for displaying token icons in 3rd party exchanges and wallets currently is a 24px ellipse. It's safe to assume that anything outside of this is at risk of not being displayed properly.

To account for this, I think we'll be pretty limited in how we choose to differentiate between an unmodified and modified token. I propose that a modified token receive the following treatments:

- A single color stroke "wrapping" the token. The color of the stroke is determined by the primary brand color of the chain where the token currently resides. (eg Ethereum lavender)
- A CSS filter to desaturate the base icon. In this way, when a token is "bridged" or "wrapped", its color is extracted, and is "returned" if it returns across the bridge.

```
.modified-token {
    border: 2px solid var(--primary-brand-color);
    filter: grayscale(100%);
}
```

While in isolation, this approach isn't as descriptive as the initial proposal, we'll need to rely on the token symbol (eg `wNEAR` or `nDAI`) to provide additional context about how the token has been modified.

Here is what this looks like in practice:

<img width="848" alt="Screen Shot 2021-03-10 at 10 19 53 AM" src="https://user-images.githubusercontent.com/5934188/110652482-6a128280-818a-11eb-84ee-6d4f7058c656.png">

**Reply by alexauroradev** [reply]
I really like that the change of the icon is CSS filtering and allows for the simple FE building.
I'm pretty OK with the general approach, though I think doing a `grayscale(100%)`, IMO, makes the icon lifeless. Having many ERC-20 tokens bridged to NEAR I don't want to end up with users on NEAR AMMs in a full despair.
Also, the primary color of the token icon is a great visual flag, so I would propose to try to conserve at least a bit of it. Perhaps, we can reduce the saturation of the icon or make it half transparent or raise the brightness?
I like the wrapping stroke.

**Reply by heycorwin** [reply]
@djsatok I agree that aesthetically it's not the best, however I think that if were to retain the color of the original icon, the stroke alone wouldn't be enough and it wouldn't be distinguishable enough from the original. An alternative is to use CSS filtering in a different way but rather than desaturating, simply using the chain's brand color as a color overlay. This looks like the following:

<img width="924" alt="Screen Shot 2021-03-11 at 10 05 42 AM" src="https://user-images.githubusercontent.com/5934188/110807747-5b42d300-8251-11eb-86c2-ff234e4ad249.png">

Perhaps this is a little less disparaging 😂

**Reply by stefanopepe** [reply]
Is there a way to filter or recognize the color code for the overlay? I like the idea of having the colors blended (blue avocados on ETH!), and I suppose we then provide sort of a "legend" that says what are the blues, the reds and any other color.

Alternatively: as a user, I want to "mouse-over" the icon to learn what every one of these icons represents, such that I can know my balance at a glimpse.

**Reply by heycorwin** [reply]
@jimmy3dita any instance that these token appear will be accompanied by its symbol (eg `nDAI`) which should provide the necessary additional context for the user to infer whatever information that the icon itself does not provide. But I agree (and @chadoh and I have discussed this) that for certain scenarios it makes sense to leverage something like a tooltip to provide more info about the token and how it has been modified.

**Reply by alexauroradev** [reply]
@corwinharrell I believe the color overlay is much better than grayscale. Let's try this out. 👍

**Reply by vgrichina** [reply]
@corwinharrell @djsatok maybe we should add a tiny ETH diamond (or 🌈 ) as a badge to refer to bridged tokens? 

exchanges can follow similar pattern or honestly do whatever, I don't think there is strict adherence to which icons to use

**Reply by vgrichina** [reply]
Also note that around 5–10% of people in US are colorblind. So relying on color coding only generally not good for accessibility. Which might be ok if we duplicate info in text, etc – just smth to keep in mind.

**Reply by chadoh** [reply]
> exchanges can follow similar pattern or honestly do whatever, I don't think there is strict adherence to which icons to use

By the end of Q2 I'd like to see us include these icons in the [on-chain metadata](https://nomicon.io/Standards/Tokens/FungibleTokenMetadata.html) to be used by all apps that display these tokens

**Reply by heycorwin** [reply]
> Also note that around 5–10% of people in US are colorblind. So relying on color coding only generally not good for accessibility. Which might be ok if we duplicate info in text, etc – just smth to keep in mind.

@vgrichina beating me at my own game! Yes thank you for this reminder. We should definitely be making sure that any implementation is accessible. I'll revisit with an accessibility check to ensure colorblind users are able to distinguish between them. The trouble with adding additional graphic elements (eg ETH diamond or the "w" as shown in the first exploration), is that it will be virtually illegible at 24px x 24px, or will be cutoff by any exchanges cropping token icons to fit the standard 24px format.
